v5/core/log.go

```diff
@@ -0,0 +1,202 @@
+package core
+
+// (C) Copyright IBM Corp. 2020, 2021.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import (
+	"log"
+	"os"
+	"sync"
+)
+
+// LogLevel defines a type for logging levels
+type LogLevel int
+
+// Log level constants
+const (
+	LevelNone LogLevel = iota
+	LevelError
+	LevelWarn
+	LevelInfo
+	LevelDebug
+)
+
+// Logger is the logging interface implemented and used by the Go core library.
+// Users of the library can supply their own implementation by calling SetLogger().
+type Logger interface {
+	Log(level LogLevel, format string, inserts ...interface{})
+	Error(format string, inserts ...interface{})
+	Warn(format string, inserts ...interface{})
+	Info(format string, inserts ...interface{})
+	Debug(format string, inserts ...interface{})
+
+	SetLogLevel(level LogLevel)
+	GetLogLevel() LogLevel
+
+	IsLogLevelEnabled(level LogLevel) bool
+	IsErrorEnabled() bool
+	IsWarnEnabled() bool
+	IsInfoEnabled() bool
+	IsDebugEnabled() bool
+}
+
+// SDKLoggerImpl is the Go core's implementation of the Logger interface.
+// This logger contains two instances of Go's log.Logger interface which are
+// used to perform message logging.  One instance represents "stdout" and other
+// represents "stderr".  Messages logged at level Error or Debug will appear on
+// the "stderr" logger, while all other messages will appear on the "stdout" logger.
```

Comment:
I suppose this is fine.  In the event that a user wants to see Error and Debug messages interleaved with Warning and Info messages, then that user just needs to pipe both stdout and stderr to the same output.

But it does seem a bit arbitrary how we've chosen to map severities to these two streams.  I mean, I could easily imagine that some users might want:
- Error and Warn -> stderr
- Info and Debug -> stdout

and other mappings are possible.

Again -- this is fine.  But if there was a way to let the user choose how to map severities to streams, without over-complicating things, it might be worth considering.

Reply:
I used this mapping of loglevel -> stream to accommodate the CLI plugin situation where we don't want anything written to stdout under normal situations because that's where the response objects are written.
However, typing that previous sentence caused me to re-think that strategy so I'll need to discuss with @dpopp07 again :)

Reply:
I discussed this with Dustin and I changed the SDKLoggerImpl type slightly to focus on the types of messages being logged rather than the particular stream.   So now we have "infoLogger" and "errorLogger".  Those will map to stdout and stderr respectively, but can easily be changed when calling NewLogger().  For example, Dustin will initialize a Logger instance that uses stderr for both sets of messages in order to avoid having logged messages interfere with operation responses inthe CLI environment.

Reply:
@padamstx I think this comment needs to change with the updated implementation, right?